core/src/main/java/cucumber/runner/OrderedTestRunEventBus.java

```diff
@@ -27,39 +37,74 @@ public Long getTime() {
         return parent.getTime();
     }
 
+    /**
+     * {@inheritDoc}
+     *
+     * @throws CucumberException when {@link TestCaseFinished} received but no {@link FeatureTestCaseExpectationEvent} was received prior
+     */
     public void send(final Event event) {
         super.send(event);
-        queue(event);
 
-        if (event instanceof TestRunFinished) {
-            parent.sendAll(flatten(queue));
-            queue.clear();
+        if (event instanceof TestCaseEvent) {
+            handleTestCaseEvent((TestCaseEvent) event);
+        }
+        else {
+            if (event instanceof FeatureTestCaseExpectationEvent) {
+                final FeatureTestCaseExpectationEvent expectation = (FeatureTestCaseExpectationEvent) event;
+                featureTestExpectation.put(expectation.getUri(), new FeatureTestCount(expectation.getExpectedCases()));
+            }
+            parent.send(event);
```

Comment:
On receiving `TestRunFinished` we could check to see if anything remains in the queue and if so then throw a `CucumberException` as it should be empty at this point